bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/hbasesnapshots/EndToEndIT.java

```diff
@@ -0,0 +1,198 @@
+ * -Dgoogle.dataflow.stagingLocation=gs://<test_bucket>/staging \
+ * -Dcloud.test.data.folder=gs://<test_bucket>/integration-test/
+ */
+public class EndToEndIT {
+
+  private static final String TEST_SNAPSHOT_NAME = "test-snapshot";
+  // Location of test data hosted on Google Cloud Storage, for on-cloud dataflow tests.
+  private static final String CLOUD_TEST_DATA_FOLDER = "cloud.test.data.folder";
+  private static final String DATAFLOW_REGION = "region";
+
+  // Column family name used in all test bigtables.
+  private static final String CF = "cf";
+
+  // Full path of the Cloud Storage folder where dataflow jars are uploaded to.
+  private static final String GOOGLE_DATAFLOW_STAGING_LOCATION = "google.dataflow.stagingLocation";
+
+  private Connection connection;
+  private String projectId;
+  private String instanceId;
+  private String tableId;
+  private String region;
+
+  private GcsUtil gcsUtil;
+  private String dataflowStagingLocation;
+  private String workDir;
+  private byte[][] keySplits;
+
+  // Snapshot data setup
+  private String hbaseSnapshotDir;
+  private String restoreDir;
+
+  @Before
+  public void setup() throws Exception {
+    projectId = getTestProperty(BigtableOptionsFactory.PROJECT_ID_KEY);
+    instanceId = getTestProperty(BigtableOptionsFactory.INSTANCE_ID_KEY);
+    dataflowStagingLocation = getTestProperty(GOOGLE_DATAFLOW_STAGING_LOCATION);
+    region = getTestProperty(DATAFLOW_REGION);
+    String cloudTestDataFolder = getTestProperty(CLOUD_TEST_DATA_FOLDER);
+    if (!cloudTestDataFolder.endsWith(File.separator)) {
+      cloudTestDataFolder = cloudTestDataFolder + File.separator;
+    }
+
+    hbaseSnapshotDir = cloudTestDataFolder + "data/";
+    UUID test_uuid = UUID.randomUUID();
+    restoreDir = cloudTestDataFolder + "restore/" + test_uuid;
+
+    // Cloud Storage config
+    GcpOptions gcpOptions = PipelineOptionsFactory.create().as(GcpOptions.class);
+    gcpOptions.setProject(projectId);
+    gcsUtil = new GcsUtil.GcsUtilFactory().create(gcpOptions);
+
+    // Bigtable config
+    connection = BigtableConfiguration.connect(projectId, instanceId);
+    tableId = "test_" + UUID.randomUUID().toString();
+
+    System.out.println("Setting up integration tests");
+
+    String[] keys = new String[] {"1", "2", "3", "4", "5", "6", "7", "8", "9"};
+    keySplits = new byte[keys.length][];
+    for (int i = 0; i < keys.length; i++) {
+      keySplits[i] = keys[i].getBytes();
+    }
+  }
+
+  private static String getTestProperty(String name) {
+    return checkNotNull(System.getProperty(name), "Required property missing: " + name);
+  }
+
+  @After
+  public void teardown() throws IOException {
+    final List<GcsPath> paths = gcsUtil.expand(GcsPath.fromUri(restoreDir + "/*"));
+
+    if (!paths.isEmpty()) {
+      final List<String> pathStrs = new ArrayList<>();
+
+      for (GcsPath path : paths) {
+        pathStrs.add(path.toString());
+      }
+      this.gcsUtil.remove(pathStrs);
```

Comment:
can you avoid all of this by just setting a ttl on the files when creating them? It will make clean up more robust

Reply:
This was the practice I found and followed in https://github.com/googleapis/java-bigtable-hbase/blob/486589b9361bf0e62b466b332bbc23964ffe4d0c/bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/sequencefiles/EndToEndIT.java#L105

If we use a TTL, since this folder could be reused across multiple consecutive test runs, it could introduce flakiness caused by timing, no?

Reply:
the workdir should always be unique due to the uuid. I think the practice in the sequencefile test can be improved and has caused issues in the past. When I wrote it I didnt think to use gcs ttls

Reply:
After some investigation, there are one more concern for using TTL:
TTL(lifecycle management for GCS) could only apply to Bucket level, to leverage that, we will need to create a separate bucket just for the restoreDir every time since current bucket also include test input files which shouldn't be deleted.

we can set up the bucket to delete its content after 1 day but the bucket will be there, so overtime the project we use for testing will accumulate a lot of empty buckets.

Current solution also has this issue(generates a lot of empty dirs instead), but this could be solved by adding:

`gcsUtil.remove(Arrays.asList(restoreDir));` after removing all files in the directory.

Reply:
Also, per this weird check: https://github.com/apache/hbase/blob/3cc24683001998f995d22591ce8d1c74f67fb9f7/hbase-server/src/main/java/org/apache/hadoop/hbase/snapshot/RestoreSnapshotHelper.java#L837

HBase treats the bucket name as part of file system so we could not possibly use another bucket for the restore...